### PR TITLE
added test cases to common_test.go and sketch_series_test.go

### PR DESCRIPTION
**Pull request by LawrenceLin690**
### What does this PR do?

Added a nil case to common_test.go for StructToMap and refactored to a table test, added zero test case for sketch_series_test.go

### CODE REVIEW

File: pkg/util/common/common_test.go [hunk 1]
```diff
@@ -128,5 +131,27 @@ func TestStructToMap(t *testing.T) {
...
+		"mymap": map[string]interface{}{},
+	}
+
+	cases := []struct {
+		name string
+		in   Top
+		out  map[string]interface{}
+	}{
+		{"notNil", top, topOutput},
+		{"nil", topNil, topNilOutput},
```
Comment (xornivore) [inline pkg/util/common/common_test.go:148]: So this `topNil` is not `nil` actually, just an instance with default values. You need to pass actual `nil` here.

File: pkg/util/common/common_test.go [hunk 2]
```diff
@@ -93,7 +94,9 @@ func TestStructToMap(t *testing.T) {
 		},
 	}
 
-	assert.Equal(t, map[string]interface{}{
+	var topNil Top
+
+	topOutput := map[string]interface{}{
```
Comment (xornivore) [inline pkg/util/common/common_test.go:99]: Could you please inline these outputs into the table below, it will be easier to follow that way.

File: pkg/util/common/common_test.go [hunk 3]
```diff
@@ -128,5 +131,27 @@ func TestStructToMap(t *testing.T) {
...
+		name string
+		in   Top
+		out  map[string]interface{}
+	}{
+		{"notNil", top, topOutput},
+		{"nil", topNil, topNilOutput},
+	}
+
+	for _, tc := range cases {
+		t.Run(fmt.Sprintf("%v", tc.name), func(t *testing.T) {
```
Comment (xornivore) [inline pkg/util/common/common_test.go:152]: No need for `Sprintf` here - can just use `tc.name`.

File: pkg/util/common/common_test.go (hunk 2 above)
Comment (LawrenceLin690) [reply pkg/util/common/common_test.go:99]: will do, thanks

File: pkg/util/common/common_test.go (hunk 3 above)
Comment (LawrenceLin690) [reply pkg/util/common/common_test.go:152]: will do, thanks

File: pkg/util/common/common_test.go (hunk 1 above)
Comment (LawrenceLin690) [reply pkg/util/common/common_test.go:148]: Whats the right way to pass in nil here? 
Looking at the code in common.go, do we want to hit the first conditional:
```
	rt, rv := reflect.TypeOf(obj), reflect.ValueOf(obj)
	if rt != nil && rt.Kind() != reflect.Struct {
		return make(map[string]interface{}, 0)
	}
```
The only instances I can get to show a <nil> type are error and an empty interface or an interface where both the type and value are nil.  However, these throw errors when passed into the original func StructToMap.  I've also passed in a nil pointers , which evaluate as expected, but I don't think they satisfy being nil here either.

Comment (xornivore) [reply pkg/util/common/common_test.go:148]: That's the point, if you see `StructToMap` panic on a particular value passed in, it's a bug that needs to be fixed. It should be resilient to these types of inputs.

File: pkg/util/common/common.go [hunk 4]
```diff
@@ -39,6 +39,15 @@ func (s StringSet) GetAll() []string {
 // StructToMap converts a struct to a map[string]interface{} based on `json` annotations defaulting to field names
 func StructToMap(obj interface{}) map[string]interface{} {
 	rt, rv := reflect.TypeOf(obj), reflect.ValueOf(obj)
+
+	if rt == nil {
+		return nil
+	}
```
Comment (LawrenceLin690) [inline pkg/util/common/common.go:45]: made a fairly simple addition here if a nil interface or error is passed in.  Returning nil will actually return map[string]interface{}(nil) in line with the specified return type.  Let me know if there's a better way to handle this, debated adding an error return type but then downstream functions will have to break out the returns.

Comment (xornivore) [reply pkg/util/common/common.go:45]: Could also return an empty map. @vboulineau have an opinion?

File: pkg/util/common/common_test.go [hunk 5]
```diff
@@ -93,26 +93,36 @@ func TestStructToMap(t *testing.T) {
...
+
+	numCase := 1
+	letterCase := "a"
+
+	cases := []struct {
+		name string
+		in   interface{}
+		out  map[string]interface{}
+	}{
+		{"nil interface ", nilCase, map[string]interface{}(nil)},
```
Comment (xornivore) [inline pkg/util/common/common_test.go:119]: I would inline inputs in the table here as well, otherwise it's hard to follow. Also - it's better to specify field names in struct initialization, that way if the order in the struct changes it's not an issue.

File: pkg/util/common/common_test.go [hunk 6]
```diff
@@ -93,26 +93,36 @@ func TestStructToMap(t *testing.T) {
...
+		{"number ", numCase, map[string]interface{}{}},
```
Comment (xornivore) [inline pkg/util/common/common_test.go:120]: This should be called "integer".

File: pkg/util/common/common_test.go [hunk 7]
```diff
@@ -93,26 +93,36 @@ func TestStructToMap(t *testing.T) {
...
+		{"letter ", letterCase, map[string]interface{}{}},
```
Comment (xornivore) [inline pkg/util/common/common_test.go:121]: This should be called "string".

File: pkg/util/common/common_test.go [hunk 8]
```diff
@@ -127,6 +137,33 @@ func TestStructToMap(t *testing.T) {
...
+						map[string]interface{}{
+							"name":        "ms2",
+							"value":       float64(2),
+							"JSONLessStr": "",
+						},
+					},
+				},
+			},
+		}},
+		{"sliceOfStructs", topSliceOfStructs, map[string]interface{}{}},
```
Comment (xornivore) [inline pkg/util/common/common_test.go:158]: You can keep spaces in the test case names for consistency - `t.Run` handles that.

**Review by xornivore — COMMENTED** [review]
Had few comments, looking good overall.

File: pkg/util/common/common_test.go (hunk 6 above)
Comment (LawrenceLin690) [reply pkg/util/common/common_test.go:120]: ✔️

File: pkg/util/common/common_test.go (hunk 7 above)
Comment (LawrenceLin690) [reply pkg/util/common/common_test.go:121]: ✔️

File: pkg/util/common/common_test.go (hunk 8 above)
Comment (LawrenceLin690) [reply pkg/util/common/common_test.go:158]: ✔️

File: pkg/util/common/common.go (hunk 4 above)
Comment (xornivore) [reply pkg/util/common/common.go:45]: I meant return `map[string]interface{}{}` for the `if rt == nil` case.

File: pkg/util/common/common_test.go (hunk 5 above)
Comment (LawrenceLin690) [reply pkg/util/common/common_test.go]: ✔️

File: pkg/util/common/common.go (hunk 4 above)
Comment (LawrenceLin690) [reply pkg/util/common/common.go]: done (changed to `map[string]interface{}{} `for the `nil` case)

File: pkg/util/common/common_test.go [hunk 9]
```diff
@@ -93,40 +93,133 @@ func TestStructToMap(t *testing.T) {
...
+		},
+		{
+			testName: "nil pointer",
+			input:    nilPointer,
+			output:   map[string]interface{}{},
+		},
+		{
+			testName: "pointer",
+			input:    &top,
+			output:   map[string]interface{}{},
```
Comment (LawrenceLin690) [inline pkg/util/common/common_test.go:215]: The expected output for a pointer to a valid struct is proposed to be the same as the output for that struct.  Just not sure currently how to access the entire struct from a pointer, running into` invalid indirect of obj` errors when trying to use `*obj`.  Do I need to rebuild it or is there any other method similar to unmarshaling that can be used?

File: pkg/util/common/common_test.go [hunk 10]
```diff
@@ -93,40 +93,133 @@ func TestStructToMap(t *testing.T) {
...
+		},
+		{
+			testName: "string",
+			input:    "a",
+			output:   map[string]interface{}{},
+		},
+		{
+			testName: "default struct",
+			input:    defaultStruct,
+			output:   map[string]interface{}{},
```
Comment (LawrenceLin690) [inline pkg/util/common/common_test.go:122]: should the output for a default struct be `map[string]interface{}{}` or something more specific?

File: pkg/util/common/common_test.go (hunk 9 above)
Comment (LawrenceLin690) [reply pkg/util/common/common_test.go]: done with: 
```
	if rt.Kind() == reflect.Ptr {
		obj = rv.Elem().Interface()
		rt, rv = reflect.TypeOf(obj), reflect.ValueOf(obj)
	}
```

File: pkg/metrics/sketch_series_test.go [hunk 11]
```diff
@@ -120,3 +120,20 @@ func TestSketchSeriesListJSONMarshal(t *testing.T) {
...
 }
+
+func TestSketchSeriesListJSONMarshalZero(t *testing.T) {
+	slZero := make(SketchSeriesList, 0)
+
+	for i := range slZero {
+		slZero[i] = makeseries(i)
+	}
```
Comment (djmitche) [inline pkg/metrics/sketch_series_test.go:129]: I believe this is redundant since the slice is empty

File: pkg/util/common/common.go [hunk 12]
```diff
@@ -39,36 +39,45 @@ func (s StringSet) GetAll() []string {
...
+	if rt == nil || rv.IsZero() {
+		return out
 	}
 
-	out := make(map[string]interface{}, rt.NumField())
-	for i := 0; i < rt.NumField(); i++ {
-		field := rt.Field(i)
+	if rt.Kind() == reflect.Ptr {
+		obj = rv.Elem().Interface()
+		rt, rv = reflect.TypeOf(obj), reflect.ValueOf(obj)
```
Comment (djmitche) [inline pkg/util/common/common.go:51]: Did you want to recurse here, to handle the `rt == nil` and `rt.Kind() == reflect.Ptr` cases?

**Review by djmitche — COMMENTED** [review]
It looks like this changes production code, in addition to the added/refactored tests mentioned in the PR description.  Can you update the PR to indicate what changes were made and why?
